core/unit_test/TestTeamScan.hpp

```diff
@@ -0,0 +1,141 @@
+  using view_type       = Kokkos::View<value_type**, execution_space>;
+
+  view_type a_d;
+  view_type a_r;
+  int32_t M = 0;
+  int32_t N = 0;
+
+  KOKKOS_FUNCTION
+  void operator()(const member_type& team) const {
+    auto leagueRank = team.league_rank();
+
+    auto beg = 0;
+    auto end = N;
+
+    Kokkos::parallel_for(
+        Kokkos::TeamThreadRange(team, beg, end),
+        [&](const int i) { a_d(leagueRank, i) = leagueRank * N + i; });
+
+    Kokkos::parallel_scan(Kokkos::TeamThreadRange(team, beg, end),
+                          [&](int i, int32_t& val, const bool final) {
+                            val += a_d(leagueRank, i);
+                            if (final) a_r(leagueRank, i) = val;
+                          });
+  }
+
+  auto operator()(int32_t _M, int32_t _N) {
+    M   = _M;
+    N   = _N;
+    a_d = view_type("a_d", M, N);
+    a_r = view_type("a_r", M, N);
+
+    Kokkos::parallel_for(policy_type(M, Kokkos::AUTO()), *this);
+
+    auto a_i = Kokkos::create_mirror_view(a_d);
+    auto a_o = Kokkos::create_mirror_view(a_r);
+    Kokkos::deep_copy(a_i, a_d);
+    Kokkos::deep_copy(a_o, a_r);
+
+    for (int32_t i = 0; i < M; ++i) {
+      value_type _scan_real = 0;
+      value_type _scan_calc = 0;
+      value_type _epsilon   = std::numeric_limits<value_type>::epsilon();
+      // each fp addition is subject to small loses in precision and these
+      // compound as loop so we set the base error to be the machine epsilon and
+      // then add in another epsilon each iteration. For example, with CUDA
+      // backend + 32-bit float + large N values (e.g. 1,000) + high
+      // thread-counts (e.g. 1024), this test will fail w/o epsilon
+      // accommodation
+      for (int32_t j = 0; j < N; ++j) {
+        _scan_real += a_i(i, j);
+        _scan_calc     = a_o(i, j);
+        auto _get_mesg = [=]() {
+          std::stringstream ss, idx;
+          idx << "(" << i << ", " << j << ") = ";
+          ss << "a_d" << idx.str() << a_i(i, j);
+          ss << ", a_r" << idx.str() << a_o(i, j);
+          return ss.str();
+        };
+        if (std::is_integral<value_type>::value) {
+          ASSERT_EQ(_scan_real, _scan_calc) << _get_mesg();
+        } else {
+          _epsilon += std::numeric_limits<value_type>::epsilon();
+          ASSERT_NEAR(_scan_real, _scan_calc, _epsilon) << _get_mesg();
+        }
+      }
+    }
+  }
+};
+
+TEST(TEST_CATEGORY, team_scan) {
+  TestTeamScan<TEST_EXECSPACE, int16_t>{}(99, 32);
+  TestTeamScan<TEST_EXECSPACE, uint16_t>{}(139, 64);
+  TestTeamScan<TEST_EXECSPACE, int32_t>{}(163, 128);
+  TestTeamScan<TEST_EXECSPACE, int64_t>{}(2156, 512);
+  TestTeamScan<TEST_EXECSPACE, uint64_t>{}(1234, 1024);
+  TestTeamScan<TEST_EXECSPACE, float>{}(108, 16);
+  TestTeamScan<TEST_EXECSPACE, float>{}(152, 80);
+  TestTeamScan<TEST_EXECSPACE, double>{}(34, 32);
+  TestTeamScan<TEST_EXECSPACE, double>{}(956, 128);
+  TestTeamScan<TEST_EXECSPACE, double>{}(2596, 512);
```

Comment:
Please add a few tests where the second argument is not a multiple of 16 and one where it is zero.

Reply:
Done.